fix: Socket leak on error

stevenh opened the pull request:
Fix sockets being leaked if we get any errors by calling agent.shutdown to close all active connections.

Improve connection handling adding support for backoff and connection termination on logstash shutdown.

New configuration options added:
* retry_initial
* retry_max
* retry_reset
* debug_status

Also:
* Remove unused socket require.
* Clean up logging.
* Bump version to 4.0.5

Fixes #20 
Fixes #21

stevenh committed:
fix: Socket leak on error

Fix sockets being leaked if we get any errors by calling agent.shutdown
to close all active connections.

Also:
* Add retry configuration parameter.

Fixes #20

stevenh commented:
In the hope it will be helpful for others I had extreme difficulty getting to the stage I could run the rspec tests for this repo so here's some hints:
1.  ubuntu 18 LTS jruby in apt is broken don't us it, you need `9.2` or above otherwise you will just get an error trying to run `jruby -S bundle install`
```
/usr/lib/ruby/vendor_ruby/psych.rb:6: warning: already initialized constant SNAKEYAML_VERSION
/usr/lib/ruby/vendor_ruby/psych.rb:6: warning: already initialized constant ANY
/usr/lib/ruby/vendor_ruby/psych.rb:6: warning: already initialized constant UTF8
/usr/lib/ruby/vendor_ruby/psych.rb:6: warning: already initialized constant UTF16LE
/usr/lib/ruby/vendor_ruby/psych.rb:6: warning: already initialized constant UTF16BE
/usr/share/jruby/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:55:in `require':
It seems your ruby installation is missing psych (for YAML output).
To eliminate this warning, please install libyaml and reinstall your ruby.
LoadError: load error: psych -- java.lang.NoClassDefFoundError: Could not initialize class org.jruby.ext.psych.PsychEmitter
           require at org/jruby/RubyKernel.java:956
           require at /usr/share/jruby/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:55
            <main> at /usr/share/jruby/lib/ruby/stdlib/yaml.rb:6
           require at org/jruby/RubyKernel.java:956
           require at /usr/share/jruby/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:55
            (root) at /usr/share/jruby/lib/ruby/stdlib/rubygems.rb:1
         load_yaml at /usr/share/jruby/lib/ruby/stdlib/rubygems.rb:701
         load_file at /usr/share/jruby/lib/ruby/stdlib/rubygems/config_file.rb:340
        initialize at /usr/share/jruby/lib/ruby/stdlib/rubygems/config_file.rb:202
  do_configuration at /usr/share/jruby/lib/ruby/stdlib/rubygems/gem_runner.rb:75
            <main> at /usr/share/jruby/lib/ruby/stdlib/rubygems/gem_runner.rb:40
              load at org/jruby/RubyKernel.java:974
            <main> at /usr/share/jruby/bin/gem:4
```
The latest version of jruby can be downloaded here:
[https://www.jruby.org/download](https://www.jruby.org/download)

2. This repo uses a logstash 5.6 which only supports java 8, if you use a later version you'll get the following error.
```
An error occurred while loading ./spec/inputs/websocket_spec.rb.
Failure/Error: require "logstash/devutils/rspec/spec_helper"

NameError:
  missing class name (`org.logstash.common.DeadLetterQueueFactory')
# ./spec/inputs/websocket_spec.rb:1:in `<main>'
```

In our case we not only got connections in `CLOSE_WAIT` but also in `ESTABLISHED` so many it ran the machine out of file handles, breaking everything.

stevenh committed:
feat: Improve connection handling

Improve connection handling adding support for backoff and connection
termination on logstash shutdown.

New configuration options added:
* retry_initial
* retry_max
* retry_reset
* warn_404

Also:
* Remove unused socket require.
* Clean up logging.
* Bump version to 4.0.5

Fixes #21

magnusbaeck commented on lib/logstash/inputs/websocket.rb:
```diff
@@ -13,32 +12,110 @@ class LogStash::Inputs::Websocket < LogStash::Inputs::Base
+    @logger.warn("Run error ",
+                 :exception => e,
+                 :url => @url_safe,
+                 :retry => @interval) unless stop?
```
Is there any particular point with the special handling of EOFError given the subsequent catch-all rescue block?

stevenh replied:
To avoid log spam when EOF is encountered its debug not warn. In our case that significantly reduced the amount of useless logging in normal operation.

magnusbaeck replied:
Oh dear, I totally missed the debug vs. warn difference despite comparing the two calls several times. Yes, it makes sense to not spam about closed connections.

stevenh replied:
Easy to miss, hopefully I've addressed all your other feedback too, but if I did miss anything just let me know.

magnusbaeck commented on lib/logstash/inputs/websocket.rb:
```diff
@@ -13,32 +12,110 @@ class LogStash::Inputs::Websocket < LogStash::Inputs::Base
+        @logger.debug("Connect failed",
+                     :status => r.status_line,
+                     :url => @url_safe,
+                     :retry => @interval) unless stop?
```
Why should 404s be treated differently that other status codes? Under which circumstances would one set `warn_404` to false?

stevenh replied:
We have it set to false as we expect 404 in normal operation i.e. the services are only there and listening when needed. This felt like a relatively special case hence the config option, defaulted to true to warn.

magnusbaeck replied:
In that case, how about instead adding an array option that lists which response codes should be ignored? The default would probably be empty but you could set it to `[404]`. That'd be more generic and, IMHO, easier to understand than a very specific option like `warn_404`.

stevenh replied:
nice idea, will do.

magnusbaeck commented on lib/logstash/inputs/websocket.rb:
```diff
@@ -13,32 +12,110 @@ class LogStash::Inputs::Websocket < LogStash::Inputs::Base
+      @interval = @retry_max
+    else
+      @interval = @interval * 2
+    end
```
Nit: Equivalent and IMHO more idiomatic:

```ruby
else
  @interval = [@interval * 2, @retry_max].min
end
```

stevenh replied:
Thanks, its been a loooong time since I did ruby ;-)

magnusbaeck commented on lib/logstash/inputs/websocket.rb:
```diff
@@ -13,32 +12,110 @@ class LogStash::Inputs::Websocket < LogStash::Inputs::Base
+
+  # The number of log entries that have to be processed before
+  # the retry interval is reset to retry_initial.
+  config :retry_reset, :validate => :number, :default => 20
```
Resetting the retry interval to the initial value after a certain number of iterations is not a pattern I have seen before. What's the rationale?

I don't think the description of the option is clear to people. Suggestion: "The number of errors that can occur before the retry interval is reset to `retry_interval`."

stevenh replied:
The backoff is there to ensure the service we're connecting to has chance to recover, however we don't want backoff to be permanent. Once we've successfully reconnected and processed some entries (`retry_rest`) the `interval` gets reset so we can quickly recover from a small interruptions.

magnusbaeck replied:
I see. But why not just reset the backoff interval as soon as we've reconnected and processed a single message?

stevenh replied:
If the service has an issue it may be able to connect and process a single message before failing again, so wanted a little more control over it than that.

Happy to have this default to 1, if you think that's better?

magnusbaeck replied:
Okay. Let's keep it configurable but default to one.

magnusbaeck reviewed (changes requested):
Apart from the inline comments that I'd like to talk about, the documentation in [master/docs/index.asciidoc](https://github.com/logstash-plugins/logstash-input-websocket/blob/master/docs/index.asciidoc) also needs to be updated. Previously the plugin documentation was generated from the comments in the code but nowadays it's maintained in separate files.

(I mistakenly added a couple of comments as single comments. They should've been part of the review.)

stevenh committed:
chore: Use idiomatic min in backoff

Use more idiomatic code to limit @interval in backoff to @retry_max.

Also:
* Clarify the use / reason for retry_reset.

stevenh committed:
chore: Add new configuration options to docs

Add the new configuration options to the ascii docs.

magnusbaeck commented on docs/index.asciidoc:
```diff
@@ -33,6 +33,10 @@ This plugin supports the following configuration options plus the <<plugins-{typ
+| <<plugins-{type}s-{plugin}-retry_initial>> |<<number,number>>|No
+| <<plugins-{type}s-{plugin}-retry_max>> |<<number,number>>|No
+| <<plugins-{type}s-{plugin}-retry_reset>> |<<number,number>>|No
+| <<plugins-{type}s-{plugin}-warn_404>> |<<boolean,boolean>>|No
```
Maintain alphabetical order of options to make navigation easier?

magnusbaeck commented on lib/logstash/inputs/websocket.rb:
```diff
@@ -13,32 +12,111 @@ class LogStash::Inputs::Websocket < LogStash::Inputs::Base
+  config :retry_reset, :validate => :number, :default => 20
+
+  # Logs 404 responses as warnings if true otherwise as debug.
+  config :warn_404, :validated => :boolean, :default => true
```
s/validated/validate/

magnusbaeck reviewed (changes requested):
Sorry for the slow review. I found a few documentation nits, not terribly important but trivial to address, and I'm still not convinced by some of the proposed configuration options.

stevenh committed:
chore: Improve config options

Improve the configuration options:
* Switch from warn_404 to debug_status.
* Change default for retry_reset to 1 so a single processed message
  will reset the retry interval.

Also:
* Correct typos in docs.
* Order options in docs so they are easier to find.

magnusbaeck commented on lib/logstash/inputs/websocket.rb:
```diff
@@ -13,32 +12,112 @@ class LogStash::Inputs::Websocket < LogStash::Inputs::Base
+
+  # Logs responses with the given HTTP status codes as debug instead
+  # of warning.
+  config :debug_status, :default => [404]
```
Add `:validate => :array`? (And change default value to empty array as per other discussion thread.)

magnusbaeck commented on lib/logstash/inputs/websocket.rb:
```diff
@@ -13,32 +12,112 @@ class LogStash::Inputs::Websocket < LogStash::Inputs::Base
+    require "uri"
+
+    p = URI.parse(@url)
+    if p.userinfo != ""
```
FWICT p.userinfo is nil if no credentials are provided so this should rather be:

    unless p.userinfo.nil?

As it stands we're adding "***:***" to the URL even if no credentials are provided which makes the log message confusing.

magnusbaeck reviewed (changes requested):
Sorry, just a few more things.

stevenh committed:
fix: userinfo cleanup

Fix userinfo cleanup which is nil not empty string when no user
credentials are present.

Also:
* Default debug_status to empty so users can easily detect invalid
  url configuration option.

stevenh commented:
@magnusbaeck does this address all your concerns?

magnusbaeck commented on lib/logstash/inputs/websocket.rb:
```diff
@@ -27,7 +27,7 @@ class LogStash::Inputs::Websocket < LogStash::Inputs::Base
   # Logs responses with the given HTTP status codes as debug instead
   # of warning.
-  config :debug_status, :default => [404]
+  config :debug_status, :valdate => :array, :default => []
```
s/valdate/validate/

stevenh replied:
Doh, fixed

magnusbaeck commented:
Sorry, I've been very busy with other things this week. There's a small but significant typo in the last commit but otherwise we're good.

stevenh committed:
fix: Typo in debug_status validate

Fix a typo in the definition of debug_status.

stevenh commented:
No problem, totally appreciate that.

Typo fixed LMK if there's anything else as keen to get this closed out.

magnusbaeck commented:
Merged. Thanks for your patience! Unfortunately it turns out I lack permissions to publish the gem to rubygems.org, but I hope that'll get resolved during Monday.
